File: vi/codes/cpp/chapter_searching/hashing_search.cpp
/**
 * File: hashing_search.cpp
 * Created Time: 2022-11-25
 * Author: krahets (krahets@163.com)
 */

#include "../utils/common.hpp"

/* Tìm kiếm băm (mảng) */
int hashingSearchArray(unordered_map<int, int> map, int target) {
    // Khóa của bảng băm: phần tử mục tiêu, giá trị: chỉ số
    // Nếu khóa này không tồn tại trong bảng băm, trả về -1
    if (map.find(target) == map.end())
        return -1;
    return map[target];
}

/* Tìm kiếm băm (danh sách liên kết) */
ListNode *hashingSearchLinkedList(unordered_map<int, ListNode *> map, int target) {
    // Khóa của bảng băm: giá trị nút mục tiêu, giá trị: đối tượng nút
    // Trả về nullptr nếu khóa không tồn tại trong bảng băm
    if (map.find(target) == map.end())
        return nullptr;
    return map[target];
}

/* Mã chạy chương trình */
int main() {
    int target = 3;

    /* Tìm kiếm băm (mảng) */
    vector<int> nums = {1, 5, 3, 2, 4, 7, 5, 9, 10, 8};
    // Khởi tạo bảng băm
    unordered_map<int, int> map;
    for (int i = 0; i < nums.size(); i++) {
        map[nums[i]] = i; // khóa: phần tử, giá trị: chỉ số
    }
    int index = hashingSearchArray(map, target);
    cout << "Index of target element 3 = " << index << endl;

    /* Tìm kiếm băm (danh sách liên kết) */
    ListNode *head = vecToLinkedList(nums);
    // Khởi tạo bảng băm
    unordered_map<int, ListNode *> map1;
    while (head != nullptr) {
        map1[head->val] = head; // khóa: giá trị nút, giá trị: nút
        head = head->next;
    }
    ListNode *node = hashingSearchLinkedList(map1, target);
    cout << "Node object corresponding to target node value 3 is " << node << endl;

    return 0;
}

File: vi/codes/swift/chapter_searching/hashing_search.swift
/**
 * File: hashing_search.swift
 * Created Time: 2023-01-28
 * Author: nuomi1 (nuomi1@qq.com)
 */

import utils

/* Tìm kiếm bằng băm (mảng) */
func hashingSearchArray(map: [Int: Int], target: Int) -> Int {
    // Khóa của bảng băm: phần tử target, giá trị: chỉ số
    // Nếu khóa này không tồn tại trong bảng băm, trả về -1
    return map[target, default: -1]
}

/* Tìm kiếm bằng băm (danh sách liên kết) */
func hashingSearchLinkedList(map: [Int: ListNode], target: Int) -> ListNode? {
    // Khóa của bảng băm: giá trị nút target, giá trị: đối tượng nút
    // Nếu khóa không có trong bảng băm, trả về null
    return map[target]
}

@main
enum HashingSearch {
    /* Mã điều khiển */
    static func main() {
        let target = 3

        /* Tìm kiếm bằng băm (mảng) */
        let nums = [1, 5, 3, 2, 4, 7, 5, 9, 10, 8]
        // Khởi tạo bảng băm
        var map: [Int: Int] = [:]
        for i in nums.indices {
            map[nums[i]] = i // khóa: phần tử, giá trị: chỉ số
        }
        let index = hashingSearchArray(map: map, target: target)
        print("Chỉ số của phần tử target 3 = \(index)")

        /* Tìm kiếm bằng băm (danh sách liên kết) */
        var head = ListNode.arrToLinkedList(arr: nums)
        // Khởi tạo bảng băm
        var map1: [Int: ListNode] = [:]
        while head != nil {
            map1[head!.val] = head! // khóa: giá trị nút, giá trị: nút
            head = head?.next
        }
        let node = hashingSearchLinkedList(map: map1, target: target)
        print("Đối tượng nút tương ứng với giá trị nút target 3 là \(node!)")
    }
}
